### PR TITLE
Add SequenceType methods to AnySequence

### PR DESCRIPTION
**Pull request by therealbnut**
Implements other methods of SequenceType in AnySequence so that overrides work correctly.

Currently `AnySequence(0...1000000000).suffix(2)` takes a long time to run, but `(0...1000000000).suffix(2)` does not.

### CODE REVIEW

**Comment by therealbnut** [comment]
See https://bugs.swift.org/browse/SR-413 for more details.

File: stdlib/public/core/ExistentialCollection.swift.gyb [hunk 1]
```diff
@@ -132,6 +132,28 @@ internal class _AnySequenceBox<Element> {
 
   internal func _underestimateCount() -> Int { _abstract() }
 
+  internal func _map<T>(@noescape transform: Element throws -> T) rethrows -> [T] { _abstract(); }
```
Comment (gribozavr) [inline stdlib/public/core/ExistentialCollection.swift.gyb]: Please remove semicolons and wrap to 80 columns (here and everywhere in the patch).

File: stdlib/public/core/ExistentialCollection.swift.gyb [hunk 2]
```diff
@@ -132,6 +132,28 @@ internal class _AnySequenceBox<Element> {
...
+  
+  internal func _filter(@noescape includeElement: Element throws -> Bool
+  ) rethrows -> [Element] {
```
Comment (gribozavr) [inline stdlib/public/core/ExistentialCollection.swift.gyb]: Please format it like this:

```
  internal func _filter(
    @noescape includeElement: Element throws -> Bool
   ) rethrows -> [Element] {
```

File: stdlib/public/core/ExistentialCollection.swift.gyb [hunk 3]
```diff
@@ -132,6 +132,28 @@ internal class _AnySequenceBox<Element> {
...
+  
+  internal func _dropLast(n: Int) -> AnySequence<Element> { _abstract(); }
+  
+  internal func _prefix(maxLength: Int) -> AnySequence<Element> { _abstract(); }
+  
+  internal func _suffix(maxLength: Int) -> AnySequence<Element> { _abstract(); }
+  
+  internal func _split(maxSplit: Int, allowEmptySlices: Bool, 
+    @noescape isSeparator: Element throws -> Bool
+  ) rethrows -> [AnySequence<Element>] {
```
Comment (gribozavr) [inline stdlib/public/core/ExistentialCollection.swift.gyb]:
```swift
  internal func _split(
    maxSplit: Int,
    allowEmptySlices: Bool, 
    @noescape isSeparator: Element throws -> Bool
  ) rethrows -> [AnySequence<Element>] {
```

File: stdlib/public/core/ExistentialCollection.swift.gyb [hunk 4]
```diff
@@ -139,6 +161,7 @@ internal class _AnySequenceBox<Element> {
   internal func _copyToNativeArrayBuffer() -> _ContiguousArrayStorageBase {
     _abstract()
   }
+
```
Comment (gribozavr) [inline stdlib/public/core/ExistentialCollection.swift.gyb]: Please remove unrelated whitespace changes.

File: stdlib/public/core/ExistentialCollection.swift.gyb [hunk 5]
```diff
@@ -223,6 +282,8 @@ public struct AnySequence<Element> : SequenceType {
   @available(*, unavailable, renamed="Element")
   public typealias T = Element
 
+  public typealias SubSequence = AnySequence<AnySequence.Generator.Element>
```
Comment (gribozavr) [inline stdlib/public/core/ExistentialCollection.swift.gyb]: I think this is just

```swift
public typealias SubSequence = AnySequence<Element>
```

File: stdlib/public/core/ExistentialCollection.swift.gyb [hunk 6]
```diff
@@ -242,6 +303,44 @@ public struct AnySequence<Element> : SequenceType {
...
+    @noescape includeElement: Element throws -> Bool
+  ) rethrows -> [Element] {
+    return try _box._filter(includeElement)
+  }
+  
+  public func forEach(@noescape body: Element throws -> Void) rethrows {
+    try _box._forEach(body)
+  }
+
+  public func dropFirst(n: Int) -> AnySequence.SubSequence {
```
Comment (gribozavr) [inline stdlib/public/core/ExistentialCollection.swift.gyb]: Why not just `-> SubSequence`?

File: stdlib/public/core/ExistentialCollection.swift.gyb [hunk 7]
```diff
@@ -242,6 +303,44 @@ public struct AnySequence<Element> : SequenceType {
   public func generate() -> AnyGenerator<Element> {
     return _box.generate()
   }
+  
+  public func map<T>(
```
Comment (gribozavr) [inline stdlib/public/core/ExistentialCollection.swift.gyb]: I think you need to copy comments from the protocol.  Swift comment parsing support does not look on other related declarations for documentation comments.

File: stdlib/public/core/ExistentialCollection.swift.gyb (hunk 3 above)
Comment (therealbnut) [inline stdlib/public/core/ExistentialCollection.swift.gyb]: Thanks, I'll fix this. FYI I copied this formatting from split in [Sequence.swift](https://github.com/apple/swift/blob/master/stdlib/public/core/Sequence.swift#L180)

File: stdlib/public/core/ExistentialCollection.swift.gyb (hunk 5 above)
Comment (therealbnut) [inline stdlib/public/core/ExistentialCollection.swift.gyb]: I had that initially but it was ambiguous with the `SequenceType` protocol extension implementation.

**Comment by therealbnut** [comment]
Thanks @gribozavr, I've applied your feedback, just verifying the tests pass locally.

File: stdlib/public/core/ExistentialCollection.swift.gyb [hunk 8]
```diff
@@ -164,6 +195,51 @@ internal class _${Kind}Box<S : ${Kind}Type>
...
+  }
+  
+  override func _forEach(
+    @noescape body: S.Generator.Element throws -> Void
+  ) rethrows {
+    return try _base.forEach(body)
+  }
+  
+  override func _dropFirst(n: Int) -> AnySequence<Element> {
+    return _base.dropFirst(n)
```
Comment (gribozavr) [inline stdlib/public/core/ExistentialCollection.swift.gyb]: I don't think this actually forwards to the underlying sequence.  It just forces the overload resolution to pick up the default implementation.

**Comment by gribozavr** [comment]
@therealbnut Thank you.  I don't think this change actually achieves what you wanted.  Could you take a look at #220?

Another thing that is missing is the same optimization for `Any*Collection`.

And of course, tests are missing.  The tests should use `LoggingRangeReplaceableCollection` wrapped in `AnySequence` and `Any*Collection` and verify that the implementations actually forward to the underlying collection.

**Comment by therealbnut** [comment]
Thanks @gribozavr, I've had a look at #220, good to know. I don't think it completely overlaps with this though.

I'm fairly sure this will work, but I will implement the tests and verify. I'm finding my dev environment for swift's libraries a little cumbersome at the moment.

Should I be putting tests into `CheckSequenceType.swift`? Also do you know a fast way from the command-line to just build what has changed (and run that test)?

**Comment by gribozavr** [comment]
@therealbnut The reason why this won't work is because at the point where you call, say, `base.prefix()` you are using the static return type `AnySequence` -- that is only the case in the default implementation.  The custom default implementation uses `Base.SubSequence` as the return type.  If you use that, you need wrapping into another box, and then the rest of #220.

The forwarding tests should go into `./test/1_stdlib/ExistentialCollection.swift` and should use `expectCustomizable`.

If you want to add more in-depth tests, take a look at the `validation-test/stdlib/Slice/` directory.  Those tests create a bunch of different collection arrangements, and call one of the `check*Collection` functions, that call every possible method on the collection and verify behavior correctness.  This won't show that the optimization is in the effect, since the default implementation is conservatively correct, but it will show that the type overall behaves correctly.

**Comment by therealbnut** [comment]
Ah right, I understand what you meant now - I agree. I'll add the tests and see if I can get it working, and/or put this on hold until #220.

**Comment by gribozavr** [comment]
@therealbnut JFYI: #895 has landed.

**Comment by therealbnut** [comment]
Thanks @gribozavr, and thank you for your help :) I'll see what I can get done, although I'm back at work with no free time for about 3 months.

I will probably do SequenceType first as that can potentially land separately.

On another note, I feel like all of this should be able to be automated by the compiler. I think this effort would be better spent with a more generic solution. Although, I'm unsure of the scale of the work to get this done. Also it would require a proposal. What do you think about the short-term/long-term win work tradeoff?

 * It would get us `AnySequence`, `AnyForwardCollection`, `AnyForwardIndex`, `AnyGenerator`, etc . for free. 
 * It would prevent this sort of bug from occurring again.
 * It would put control of the implementation in the hands of the compiler (do compilers have hands?).
 * It would allow future optimisations to be applied uniformly and with a consistent implementation.
 * It removes a lot of error-prone repetitive boilerplate code.

Something like this could work:
```swift
struct AnySequence<T>: @any SequenceType {
    public typealias Element = T
    public typealias SubSequence = AnySequence<T>
}
```

The compiler could generate those method definitions. These are two downsides:
 * It likely won't copy method comments from the protocol. I think that comments should also be addressed, but as a separate issue that improves all protocols.
 * It isn't explicit that it's providing a method, so would you expect SequenceType's extension or AnySequence to provide the implementation.

Alternatively it could be clearer if you explicitly state the methods, with comments. I can't think of a good way to annotate that though, and it also adds a lot of boilerplate.

**Comment by gribozavr** [comment]
@therealbnut Automatically generating an implementation like this with type erasure might be possible, but it would be a massively large compiler feature, because of the complexity of the problem, and because there are so many choices in the way you can actually do the implementation details of `AnySequence`.  I don't think it would be implemented any time soon.

What is interesting (and practical) in this area is the interface forwarding problem, its application area is much wider than type erasure, and I think it would be more interesting to implement in the language.

**Comment by therealbnut** [comment]
@gribozavr hmm, that's a good point. I think there was proposals that talked about implementation forwarding, but if I recall correctly it was jumbled in with a lot of other features.

Having a play around with how it could be expressed it seems to require a few capabilities I don't remember those proposals discussing, namely override, public, private, computed and/or abstract.

```swift
private class _SequenceBase<E>: SequenceType {
    private var _neverUsed: AnySequence<E> { return _abstract() }
    forward SequenceType = _neverUsed
}
private final class _Sequence<S: SequenceType>: _SequenceBase<S.Generator.Element> {
    private var _implementation: S
    override forward SequenceType = _implementation
}
struct AnySequence<T>: SequenceType {
    public typealias Element = T
    public typealias SubSequence = AnySequence<T>

    private var base: _SequenceBase<T>

    public forward SequenceType = base
}
```

Anyway, I'll proceed with this PR, it's at least something that would be desirable in the interim.

**Comment by therealbnut** [comment]
FYI I believe the tests to be valid, but they will be failing until the `AnySequence` implementation is updated.

@swift-ci

**Comment by DougGregor** [comment]
This has been dormant for a long time. @dabrahams , is this a direction we want to go?

**Comment by therealbnut** [comment]
@DougGregor Sorry about that. I reported the bug and had a go at adding tests and fixing it, I got stuck and had work commitments, it was also pre Swift 3 and iirc. dependent on changes out of scope of that release.

The bug (if unresolved) is still a big Swift performance issue though, and perhaps can be fixed with some of the recent improvements to generics and extensions. I think you've been doing work around that.

https://bugs.swift.org/browse/SR-413

**Comment by dabrahams** [comment]
@DougGregor at least until such time as we have generalized existentials, yes, we'd like this to happen.

**Comment by therealbnut** [comment]
I'm closing this, the work is still necessary, but this branch is well and truly stale.
